File: 29Items-Projects/7Algorithms/2GRAPH-ALGORITHMS/python/graph_algorithms/kruskal.py
from __future__ import annotations


class DisjointSet:
    def __init__(self, vertices: list[str]) -> None:
        self.parent = {vertex: vertex for vertex in vertices}
        self.rank = {vertex: 0 for vertex in vertices}

    def find(self, vertex: str) -> str:
        if self.parent[vertex] != vertex:
            self.parent[vertex] = self.find(self.parent[vertex])
        return self.parent[vertex]

    def union(self, left: str, right: str) -> bool:
        root_left = self.find(left)
        root_right = self.find(right)
        if root_left == root_right:
            return False
        if self.rank[root_left] < self.rank[root_right]:
            self.parent[root_left] = root_right
        elif self.rank[root_left] > self.rank[root_right]:
            self.parent[root_right] = root_left
        else:
            self.parent[root_right] = root_left
            self.rank[root_left] += 1
        return True


def kruskal(vertices: list[str], edges: list[tuple[str, str, int]]) -> tuple[int, list[tuple[str, str, int]]]:
    if len(vertices) != len(set(vertices)):
        raise ValueError("vertices must be unique")
    vertex_set = set(vertices)
    for source, target, _weight in edges:
        if source not in vertex_set or target not in vertex_set:
            raise ValueError("edge references an unknown vertex")
    ds = DisjointSet(vertices)
    total = 0
    tree: list[tuple[str, str, int]] = []
    for source, target, weight in sorted(edges, key=lambda edge: edge[2]):
        if ds.union(source, target):
            total += weight
            tree.append((source, target, weight))
    if vertices and len(tree) != len(vertices) - 1:
        raise ValueError("graph is disconnected")
    return total, tree


if __name__ == "__main__":
    nodes = ["A", "B", "C", "D"]
    weighted_edges = [("A", "B", 1), ("A", "C", 4), ("B", "C", 2), ("B", "D", 5), ("C", "D", 1)]
    print("Kruskal minimum spanning tree")
    print(kruskal(nodes, weighted_edges))

File: 29Items-Projects/7Algorithms/2GRAPH-ALGORITHMS/java/src/main/java/com/example/graphalgorithms/Kruskal.java
package com.example.graphalgorithms;

import java.util.*;

public class Kruskal {
    public static class Edge {
        private final String from;
        private final String to;
        private final int weight;

        public Edge(String from, String to, int weight) {
            this.from = from;
            this.to = to;
            this.weight = weight;
        }

        public String from() { return from; }
        public String to() { return to; }
        public int weight() { return weight; }
        public String toString() { return "(" + from + " -> " + to + ", " + weight + ")"; }
    }

    public static class Result {
        private final int totalWeight;
        private final List<Edge> edges;

        public Result(int totalWeight, List<Edge> edges) {
            this.totalWeight = totalWeight;
            this.edges = edges;
        }

        public int totalWeight() { return totalWeight; }
        public List<Edge> edges() { return edges; }
        public String toString() { return "Result{totalWeight=" + totalWeight + ", edges=" + edges + "}"; }
    }

    public static Result minimumSpanningTree(List<String> vertices, List<Edge> edges) {
        if (new HashSet<>(vertices).size() != vertices.size()) {
            throw new IllegalArgumentException("vertices must be unique");
        }
        Set<String> vertexSet = new HashSet<>(vertices);
        for (Edge edge : edges) {
            if (!vertexSet.contains(edge.from()) || !vertexSet.contains(edge.to())) {
                throw new IllegalArgumentException("edge references an unknown vertex");
            }
        }
        DisjointSet ds = new DisjointSet(vertices);
        List<Edge> tree = new ArrayList<>();
        int total = 0;
        edges.stream().sorted(Comparator.comparingInt(Edge::weight)).forEach(edge -> {
            if (ds.union(edge.from(), edge.to())) {
                tree.add(edge);
            }
        });
        for (Edge edge : tree) {
            total += edge.weight();
        }
        if (!vertices.isEmpty() && tree.size() != vertices.size() - 1) {
            throw new IllegalArgumentException("graph is disconnected");
        }
        return new Result(total, tree);
    }

    private static class DisjointSet {
        private final Map<String, String> parent = new HashMap<>();
        private final Map<String, Integer> rank = new HashMap<>();

        DisjointSet(List<String> vertices) {
            for (String vertex : vertices) {
                parent.put(vertex, vertex);
                rank.put(vertex, 0);
            }
        }

        String find(String vertex) {
            if (!parent.get(vertex).equals(vertex)) {
                parent.put(vertex, find(parent.get(vertex)));
            }
            return parent.get(vertex);
        }

        boolean union(String left, String right) {
            String rootLeft = find(left);
            String rootRight = find(right);
            if (rootLeft.equals(rootRight)) {
                return false;
            }
            if (rank.get(rootLeft) < rank.get(rootRight)) {
                parent.put(rootLeft, rootRight);
            } else if (rank.get(rootLeft) > rank.get(rootRight)) {
                parent.put(rootRight, rootLeft);
            } else {
                parent.put(rootRight, rootLeft);
                rank.put(rootLeft, rank.get(rootLeft) + 1);
            }
            return true;
        }
    }

    public static void main(String[] args) {
        System.out.println("Kruskal minimum spanning tree");
        System.out.println(minimumSpanningTree(
            List.of("A", "B", "C", "D"),
            List.of(new Edge("A", "B", 1), new Edge("A", "C", 4), new Edge("B", "C", 2), new Edge("C", "D", 1))
        ));
    }
}
